src/compartment.c

```diff
@@ -639,65 +427,452 @@ comp_exec(
+{
+    int lib_fd = open(lib_name, O_RDONLY);
+    char *lib_path = NULL;
+    if (lib_fd == -1)
+    {
+        // Try to find the library in dependent paths
+        // TODO currently only $COMP_LIBRARY_PATH
+        lib_path = find_in_dir(lib_name, getenv(libs_path_env_var));
+        lib_fd = open(lib_path, O_RDONLY);
+        if (lib_fd == -1)
+        {
+            errx(1, "Error opening compartment file  %s!\n", lib_path);
+        }
+    }
+
+    // Read ELF headers
+    Elf64_Ehdr lib_ehdr;
+    do_pread(lib_fd, &lib_ehdr, sizeof(Elf64_Ehdr), 0);
+    if (lib_ehdr.e_type != ET_DYN)
+    {
+        errx(1,
+            "Error parsing `%s` - only supporting ELFs of type DYN (shared "
+            "object files)!\n",
+            lib_path);
+    }
+
+    struct LibDependency *new_lib = malloc(sizeof(struct LibDependency));
+    new_lib->lib_name = malloc(strlen(lib_name));
+    strcpy(new_lib->lib_name, lib_name);
+    if (lib_path)
+    {
+        new_lib->lib_path = malloc(strlen(lib_path));
+        strcpy(new_lib->lib_path, lib_path);
+    }
+    else
+    {
+        new_lib->lib_path = malloc(strlen(lib_name));
+        strcpy(new_lib->lib_path, lib_name);
+    }
+
+    // Initialization
+    new_lib->lib_mem_base = NULL;
+
+    new_lib->lib_segs_count = 0;
+    new_lib->lib_segs_size = 0;
+    new_lib->lib_segs = NULL;
+
+    new_lib->lib_syms_count = 0;
+    new_lib->lib_syms = NULL;
+
+    new_lib->lib_dep_count = 0;
+    new_lib->lib_dep_names = NULL;
+
+    new_lib->rela_maps_count = 0;
+    new_lib->rela_maps = NULL;
+
+    parse_lib_segs(&lib_ehdr, lib_fd, new_lib, new_comp);
+
+    // Load `.shstr` section, so we can check section names
+    Elf64_Shdr shstrtab_hdr;
+    do_pread(lib_fd, &shstrtab_hdr, sizeof(Elf64_Shdr),
+        lib_ehdr.e_shoff + lib_ehdr.e_shstrndx * sizeof(Elf64_Shdr));
+    char *shstrtab = malloc(shstrtab_hdr.sh_size);
+    do_pread(lib_fd, shstrtab, shstrtab_hdr.sh_size, shstrtab_hdr.sh_offset);
+
+    // XXX The string table is read in `strtab` as a sequence of
+    // variable-length strings. Then, symbol names are obtained by indexing at
+    // the offset where the name for that symbol begins. Therefore, the type
+    // `char*` for the string table makes sense.
+    //
+    // Example:
+    // -------------------------------
+    // | "foo\0" | "bar\0" | "baz\0" |
+    // -------------------------------
+    //    0123      4567      89ab
+    //
+    // Symbol table entries will have the "name" value of the three
+    // corresponding symbols as 0, 4, and 8.
+
+    // Traverse sections once to get headers for sections of interest
+    const size_t headers_of_interest_count = 3;
+    size_t found_headers = 0;
+    Elf64_Shdr curr_shdr;
+    for (size_t i = 0; i < lib_ehdr.e_shnum; ++i)
+    {
+        do_pread(lib_fd, &curr_shdr, sizeof(Elf64_Shdr),
+            lib_ehdr.e_shoff + i * sizeof(Elf64_Shdr));
+
+        if (curr_shdr.sh_type == SHT_SYMTAB)
```

Comment:
I think this is unlikely, or maybe impossible, but what if there are two sections of the same type (e.g. `SHT-SYMTAB`)?

Reply:
~~I cry~~ For now, things will break horribly. I would need to see an actual example to try to understand what would happen practically. I'll need to read and see if it's possible to produce such a binary. I think this would be beyond the current PR, however. I will at least have a read around for now to see, and then update.

Reply:
Looking at the ELF specification [^1], page 66 (Book III, page 1-2) suggests that there can only be one `SHT_SYMTAB`, `SHT_DYNSYM`, and `SHT_DYNAMIC` (at least for now). I will also assume there can only be one section named `.rela.plt`. Therefore it seems your original assumption is (currently) impossible. I added a comment saying as such in 344cce8c178e96e3e96aee014c6dee4be6745294.

[^1]: https://refspecs.linuxfoundation.org/elf/elf.pdf

Reply:
Phew!